br/pkg/checkpoint/checkpoint.go

```diff
@@ -464,6 +524,115 @@ func (r *CheckpointRunner) doFlush(ctx context.Context, meta map[string]*RangeGr
+	var (
+		p     int64 = 0
+		l     int64 = 0
+		retry int   = 0
+	)
+	errRetry := utils.WithRetry(ctx, func() error {
+		var err error
+		p, l, err = r.timer.GetTS(ctx)
+		if err != nil {
+			retry++
+			log.Info("failed to get ts", zap.Int("retry", retry), zap.Error(err))
+			return err
+		}
+
+		return nil
+	}, utils.NewPDReqBackoffer())
+
+	return p, l, errors.Trace(errRetry)
+}
+
+// flush the lock to the external storage
+func (r *CheckpointRunner) flushLock(ctx context.Context, p int64) error {
+	lock := &CheckpointLock{
+		LockId:   r.lockId,
+		LockAt:   p,
+		ExpireAt: p + lockTimeToLive.Milliseconds(),
+	}
+	log.Info("start to flush the checkpoint lock", zap.Int64("lock-at", lock.LockAt), zap.Int64("expire-at", lock.ExpireAt))
+	data, err := json.Marshal(lock)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = r.storage.WriteFile(ctx, CheckpointLockPath, data)
+	return errors.Trace(err)
+}
+
+// check whether this lock belongs to this BR
+func (r *CheckpointRunner) checkLockFile(ctx context.Context, caller string, now int64) error {
+	data, err := r.storage.ReadFile(ctx, CheckpointLockPath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	lock := &CheckpointLock{}
+	err = json.Unmarshal(data, lock)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if lock.ExpireAt >= now && lock.LockId != r.lockId {
+		return errors.New(fmt.Sprintf("Failed to %s. The existing lock will expire in %d seconds. "+
+			"There may be another BR(%d) running. If not, you can wait for the lock to expire, or delete the file `%s%s` manually.",
+			caller, (lock.ExpireAt-now)/1000, lock.LockId, strings.TrimRight(r.storage.URI(), "/"), CheckpointLockPath))
+	}
+	if lock.LockId > r.lockId {
+		return errors.New(fmt.Sprintf("Failed to %s. There are another BR(%d) running after but setting lock before this one(%d). "+
+			"Please check whether the BR is running. If not, you can retry.", caller, lock.LockId, r.lockId))
+	}
+	if lock.LockId == r.lockId {
+		log.Warn("The lock has expired.", zap.Int64("expire-at(ms)", lock.ExpireAt), zap.Int64("now(ms)", now))
+	}
+	return nil
+}
+
+// generate a new lock and flush the lock to the external storage
+func (r *CheckpointRunner) updateLock(ctx context.Context) error {
+	p, _, err := r.getTS(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err = r.checkLockFile(ctx, "update checkpoint lock", p); err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(r.flushLock(ctx, p))
+}
+
+// Attempt to initialize the lock. Need to stop the backup when there is an unexpired locks.
+func (r *CheckpointRunner) initialLock(ctx context.Context) error {
+	p, l, err := r.getTS(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	r.lockId = oracle.ComposeTS(p, l)
+	exist, err := r.storage.FileExists(ctx, CheckpointLockPath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if exist {
+		if err := r.checkLockFile(ctx, "initialize checkpoint lock", p); err != nil {
```

Comment:
Maybe we can use `errors.Annotate` here instead of pass the caller explicitly, given the `caller` field is only used for annotating the error.